build.assets/Dockerfile-centos7-fips

```diff
@@ -1,149 +1,27 @@
-    # required by libbpf
-    elfutils-libelf-devel-static \
-    git \
-    net-tools \
-    # required by boringssl
-    ninja-build \
-    # required by Teleport PAM support
-    pam-devel \
-    perl-IPC-Cmd \
-    tree \
-    # used by our Makefile
-    which \
-    zip \
-    # required by libbpf
-    zlib-static && \
-    yum clean all
-
-# As mentioned above, these packages are unsigned.
-RUN yum install --nogpgcheck -y \
-        ${DEVTOOLSET}-gcc* \
-        ${DEVTOOLSET}-make && \
-    yum clean all
-
-# Install Go.
-ARG GOLANG_VERSION
-RUN mkdir -p /opt && cd /opt && curl -fsSL https://storage.googleapis.com/golang/${GOLANG_VERSION}.linux-${BUILDARCH}.tar.gz | tar xz && \
-    mkdir -p /go/src/github.com/gravitational/teleport && \
-    chmod a+w /go && \
-    chmod a+w /var/lib && \
-    chmod a-w /
-ENV GOEXPERIMENT=boringcrypto \
-    GOPATH="/go" \
-    GOROOT="/opt/go" \
-    PATH="/opt/llvm/bin:$PATH:/opt/go/bin:/go/bin:/go/src/github.com/gravitational/teleport/build"
-
-# Install PAM module and policies for testing.
-COPY pam/ /opt/pam_teleport/
-RUN make -C /opt/pam_teleport install
-
-RUN chmod a-w /
-
-ARG RUST_VERSION
-ENV RUSTUP_HOME=/usr/local/rustup \
-     CARGO_HOME=/usr/local/cargo \
-     PATH=/usr/local/cargo/bin:$PATH \
-     RUST_VERSION=$RUST_VERSION
-
-RUN mkdir -p $RUSTUP_HOME && chmod a+w $RUSTUP_HOME && \
-    mkdir -p $CARGO_HOME/registry && chmod -R a+w $CARGO_HOME
-
-# Install Rust using the ci user, as that is the user that
-# will run builds using the Rust toolchains we install here.
-USER ci
-RUN curl --proto '=https' --tlsv1.2 -fsSL https://sh.rustup.rs | sh -s -- -y --profile minimal --default-toolchain $RUST_VERSION && \
-    rustup --version && \
-    cargo --version && \
-    rustc --version && \
-    rustup component add rustfmt clippy && \
-    rustup target add ${TARGETARCH}-unknown-linux-gnu
-
-ARG WASM_PACK_VERSION
-# Install wasm-pack for targeting WebAssembly from Rust.
-RUN cargo install wasm-pack --version ${WASM_PACK_VERSION}
-
-ARG LIBBPF_VERSION
-COPY --from=libbpf /opt/libbpf/usr /usr/libbpf-${LIBBPF_VERSION}
-
-# Download pre-built CentOS 7 assets with clang needed to build BoringSSL and BPF tools.
-COPY --from=teleport-buildbox-centos7-assets /opt/llvm /opt/llvm
-
-VOLUME ["/go/src/github.com/gravitational/teleport"]
-EXPOSE 6600 2379 2380
+ARG BUILDBOX_CENTOS7
+FROM ${BUILDBOX_CENTOS7}-${BUILDARCH}
+
+# Enable FIPS mode (BoringCrypto) for Go
+# https://github.com/golang/go/blob/master/src/crypto/internal/boring/README.md
+ENV GOEXPERIMENT=boringcrypto
+
+# Enable OpenSSL FIPS mode by default
+# https://www.openssl.org/docs/man3.0/man7/fips_module.html
+RUN echo 'config_diagnostics = 1\n\
+openssl_conf = openssl_init\n\
+\n\
+.include /usr/local/ssl/fipsmodule.cnf\n\
+\n\
+[openssl_init]\n\
+providers = provider_sect\n\
+\n\
+[provider_sect]\n\
+fips = fips_sect\n\
+base = base_sect\n\
+\n\
+[base_sect]\n\
+activate = 1\n\
+' > /usr/local/ssl/openssl.cnf
+
+ENV OPENSSL_FIPS=1
```

Comment:
Should we attempt to run some checks that both Go and OpenSSL are correctly configured, or do we check that elsewhere in the pipelines?

Reply:
We should also check if the generated binary has "boring symbols". Sounds like it's very easy to break the FIPS build and not realize it for a long time (unless we have some verification that I'm not aware of).